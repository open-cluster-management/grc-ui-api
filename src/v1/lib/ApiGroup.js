/* Copyright (c) 2020 Red Hat, Inc. */
// acm api group and version
const ApiGroup = {
  hostUrl: 'http://0.0.0.0/kubernetes/apis',
  policiesGroup: 'policy.open-cluster-management.io',
  appsGroup: 'apps.open-cluster-management.io',
  version: 'v1',
  clusterInfoGroup: 'internal.open-cluster-management.io',
  clusterAPIVersion: 'v1beta1',
};

export default ApiGroup;
