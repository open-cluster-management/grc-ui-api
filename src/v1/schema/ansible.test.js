/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

import supertest from 'supertest';
import nock from 'nock';
import server, { GRAPHQL_PATH } from '../index';
import {
  mockAnsibleJobTemplatesResponse,
  mockAnsibleSecretsResponse,
  mockSecretExistsInTargetNamespaceResponse,
  mockSecretNotExistsInTargetNamespaceResponse,
  mockRootAnsibleSecetResponse,
  mockCopiedSecetResponse,
  mockPolicyAutomationsResponse,
  mockAnsibleJobListResponse,
  mockCreatePolicyAutomationResponse,
  mockUpdatePolicyAutomationResponse,
} from '../mocks/Ansible';
import ApiGroup from '../lib/ApiGroup';

describe('Ansible Automation Resolver', () => {
  test('Correctly resolves ansible credentials', () => new Promise((done) => {
    const APIServer = nock('http://0.0.0.0/kubernetes');
    ['local-cluster', 'cluster1', 'policy-namespace', 'default', 'kube-system'].forEach((ns) => {
      APIServer.persist().get(`/api/v1/namespaces/${ns}/secrets?labelSelector=cluster.open-cluster-management.io/provider=ans`).reply(200, mockAnsibleSecretsResponse(ns));
    });
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          ansibleCredentials{
            name
            namespace
            host
            token
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly resolves Policy Automation in single ns', () => new Promise((done) => {
    const APIServer = nock('http://0.0.0.0/kubernetes');
    const ns = 'default';
    APIServer.persist().get(`/apis/${ApiGroup.policiesGroup}/v1beta1/namespaces/${ns}/policyautomations`).reply(200, mockPolicyAutomationsResponse(ns));
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          policyAutomations{
            kind
            apiVersion
            metadata {
              name
              namespace
            }
            spec
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly resolves Policy Automation in multi ns', () => new Promise((done) => {
    const APIServer = nock('http://0.0.0.0/kubernetes');
    ['local-cluster', 'cluster1', 'policy-namespace', 'default', 'kube-system'].forEach((ns) => {
      APIServer.persist().get(`/apis/${ApiGroup.policiesGroup}/v1beta1/namespaces/${ns}/policyautomations`).reply(200, mockPolicyAutomationsResponse(ns));
    });
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          policyAutomations{
            kind
            apiVersion
            metadata {
              name
              namespace
            }
            spec
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly resolves Ansible Job Templates', () => new Promise((done) => {
    nock('https://ansible-tower.com').persist().get('/api/v2/job_templates')
      .reply(200, mockAnsibleJobTemplatesResponse);
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          ansibleJobTemplates(host:"https://ansible-tower.com", token:"bW9ja3Rva2Vu"){
            name
            description
            extra_vars
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Should return root ansible credential name as the targeNamespace is same as credential namespace', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          copyAnsibleSecret(name: "toweraccess", namespace: "default", targetNamespace: "default"){
            name
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Should return existing secret name as the copied secret already exists', () => new Promise((done) => {
    const name = 'toweraccess';
    const namespace = 'default';
    const targetNamespace = 'kube-system';
    const APIServer = nock('http://0.0.0.0/kubernetes');
    APIServer.persist().get(`/api/v1/namespaces/${targetNamespace}/secrets?labelSelector=cluster.open-cluster-management.io/copiedFromSecretName=${name},cluster.open-cluster-management.io/copiedFromNamespace=${namespace}`).reply(200, mockSecretExistsInTargetNamespaceResponse);
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          copyAnsibleSecret(name: "toweraccess", namespace: "default", targetNamespace: "kube-system"){
            name
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        nock.cleanAll(); // clear mock
        done();
      });
  }));

  test('Should copy and return secret name as credential needs to be copied', () => new Promise((done) => {
    const name = 'toweraccess';
    const namespace = 'default';
    const targetNamespace = 'kube-system';
    const APIServer = nock('http://0.0.0.0/kubernetes');
    APIServer.persist().get(`/api/v1/namespaces/${targetNamespace}/secrets?labelSelector=cluster.open-cluster-management.io/copiedFromSecretName=${name},cluster.open-cluster-management.io/copiedFromNamespace=${namespace}`).reply(200, mockSecretNotExistsInTargetNamespaceResponse);
    APIServer.persist().get(`/api/v1/namespaces/${namespace}/secrets/${name}`).reply(200, mockRootAnsibleSecetResponse);
    APIServer.persist().post(`/api/v1/namespaces/${targetNamespace}/secrets`).reply(200, mockCopiedSecetResponse);
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          copyAnsibleSecret(name: "toweraccess", namespace: "default", targetNamespace: "kube-system"){
            name
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Should correctly resolve ansible automation history', () => new Promise((done) => {
    const namespace = 'default';
    const APIServer = nock('http://0.0.0.0/kubernetes');
    APIServer.persist().get(`/apis/tower.ansible.com/v1alpha1/namespaces/${namespace}/ansiblejobs`).reply(200, mockAnsibleJobListResponse);
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `{
          ansibleAutomationHistories(name: "policy-role", namespace:"default") {
            name
            namespace
            status
            started
            finished
            job
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});

test('Correctly Resolves Create Policy Automation Mutation', () => new Promise((done) => {
  const APIServer = nock('http://0.0.0.0/kubernetes');
  ['default'].forEach((namespace) => {
    APIServer.persist().post(`/apis/${ApiGroup.policiesGroup}/v1beta1/namespaces/${namespace}/policyautomations`)
      .reply(200, mockCreatePolicyAutomationResponse);
  });
  supertest(server)
    .post(GRAPHQL_PATH)
    .send({
      query: `
      mutation {
        modifyPolicyAutomation(
          poliyAutomationJSON: [{
            kind: "PolicyAutomation",
            apiVersion: "policy.open-cluster-management.io/v1alpha1",
            metadata: {
              name: "policy-grc-default-policyAutomation",
              namespace: "default",
            },
            spec: {
              policyRef: "policy-grc-111",
              eventHook: "non-compliance",
              mode: "once",
              automationDef: {
                type: "AnsibleJob",
                name: "Demo Job Template",
                secret: "grc-testing",
                extra_vars: {
                  selector: "target-cluster",
                },
              },
            },
          }],
          action: "post"
        )
      }
    `,
    })
    .end((err, res) => {
      expect(JSON.parse(res.text)).toMatchSnapshot();
      done();
    });
}));

test('Resolves Policy Automation Mutation Empty Case', () => new Promise((done) => {
  const APIServer = nock('http://0.0.0.0/kubernetes');
  ['default'].forEach((namespace) => {
    APIServer.persist().post(`/apis/${ApiGroup.policiesGroup}/v1beta1/namespaces/${namespace}/policyautomations`)
      .reply(200, mockCreatePolicyAutomationResponse);
  });
  supertest(server)
    .post(GRAPHQL_PATH)
    .send({
      query: `
      mutation {
        modifyPolicyAutomation(
          poliyAutomationJSON: [],
          action: "post"
        )
      }
    `,
    })
    .end((err, res) => {
      expect(JSON.parse(res.text)).toMatchSnapshot();
      done();
    });
}));

test('Correctly Resolves Update Policy Automation Mutation', () => new Promise((done) => {
  const APIServer = nock('http://0.0.0.0/kubernetes');
  ['default'].forEach((namespace) => {
    APIServer.persist().put(`/apis/${ApiGroup.policiesGroup}/v1beta1/namespaces/${namespace}/policyautomations/policy-grc-default-policyAutomation`)
      .reply(200, mockUpdatePolicyAutomationResponse);
  });
  supertest(server)
    .post(GRAPHQL_PATH)
    .send({
      query: `
      mutation {
        modifyPolicyAutomation(
          poliyAutomationJSON: [{
            kind: "PolicyAutomation",
            apiVersion: "policy.open-cluster-management.io/v1alpha1",
            metadata: {
              name: "policy-grc-default-policyAutomation",
              namespace: "default",
            },
            spec: {
              policyRef: "policy-grc-111",
              eventHook: "non-compliance",
              mode: "once",
              automationDef: {
                type: "AnsibleJob",
                name: "Demo Job Template",
                secret: "grc-testing",
                extra_vars: {
                  selector: "target-cluster",
                },
              },
            },
          }],
          action: "put"
        )
      }
    `,
    })
    .end((err, res) => {
      expect(JSON.parse(res.text)).toMatchSnapshot();
      done();
    });
}));
