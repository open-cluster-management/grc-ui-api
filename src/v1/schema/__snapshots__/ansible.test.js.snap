// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ansible Automation Resolver Correctly resolves Ansible Job Templates 1`] = `
Object {
  "data": Object {
    "ansibleJobTemplates": Array [
      Object {
        "description": "",
        "extra_vars": "",
        "name": "Demo Job Template",
      },
    ],
  },
}
`;

exports[`Ansible Automation Resolver Correctly resolves ansible automations in multi ns 1`] = `
Object {
  "data": Object {
    "ansibleAutomations": Array [
      Object {
        "apiVersion": "policy.open-cluster-management.io/v1alpha1",
        "kind": "PolicyAutomation",
        "metadata": Object {
          "name": "policy-grc-default-AnsibleJob",
          "namespace": "default",
        },
        "spec": Object {
          "automationDef": Object {
            "extra_vars": Object {
              "selector": "target-cluster",
            },
            "name": "Demo Job Template",
            "secret": "grc-testing",
            "type": "AnsibleJob",
          },
          "eventHook": "non-compliance",
          "mode": "once",
          "policyRef": "policy-grc-111",
        },
      },
      Object {
        "apiVersion": "policy.open-cluster-management.io/v1alpha1",
        "kind": "PolicyAutomation",
        "metadata": Object {
          "name": "policy-grc-kube-system-AnsibleJob",
          "namespace": "kube-system",
        },
        "spec": Object {
          "automationDef": Object {
            "extra_vars": Object {
              "selector": "target-cluster",
            },
            "name": "Demo Job Template",
            "secret": "grc-testing",
            "type": "AnsibleJob",
          },
          "eventHook": "non-compliance",
          "mode": "once",
          "policyRef": "policy-grc-111",
        },
      },
      Object {
        "apiVersion": "policy.open-cluster-management.io/v1alpha1",
        "kind": "PolicyAutomation",
        "metadata": Object {
          "name": "policy-grc-policy-namespace-AnsibleJob",
          "namespace": "policy-namespace",
        },
        "spec": Object {
          "automationDef": Object {
            "extra_vars": Object {
              "selector": "target-cluster",
            },
            "name": "Demo Job Template",
            "secret": "grc-testing",
            "type": "AnsibleJob",
          },
          "eventHook": "non-compliance",
          "mode": "once",
          "policyRef": "policy-grc-111",
        },
      },
      Object {
        "apiVersion": "policy.open-cluster-management.io/v1alpha1",
        "kind": "PolicyAutomation",
        "metadata": Object {
          "name": "policy-grc-cluster1-AnsibleJob",
          "namespace": "cluster1",
        },
        "spec": Object {
          "automationDef": Object {
            "extra_vars": Object {
              "selector": "target-cluster",
            },
            "name": "Demo Job Template",
            "secret": "grc-testing",
            "type": "AnsibleJob",
          },
          "eventHook": "non-compliance",
          "mode": "once",
          "policyRef": "policy-grc-111",
        },
      },
      Object {
        "apiVersion": "policy.open-cluster-management.io/v1alpha1",
        "kind": "PolicyAutomation",
        "metadata": Object {
          "name": "policy-grc-local-cluster-AnsibleJob",
          "namespace": "local-cluster",
        },
        "spec": Object {
          "automationDef": Object {
            "extra_vars": Object {
              "selector": "target-cluster",
            },
            "name": "Demo Job Template",
            "secret": "grc-testing",
            "type": "AnsibleJob",
          },
          "eventHook": "non-compliance",
          "mode": "once",
          "policyRef": "policy-grc-111",
        },
      },
    ],
  },
}
`;

exports[`Ansible Automation Resolver Correctly resolves ansible automations in single ns 1`] = `
Object {
  "data": Object {
    "ansibleAutomations": Array [
      Object {
        "apiVersion": "policy.open-cluster-management.io/v1alpha1",
        "kind": "PolicyAutomation",
        "metadata": Object {
          "name": "policy-grc-default-AnsibleJob",
          "namespace": "default",
        },
        "spec": Object {
          "automationDef": Object {
            "extra_vars": Object {
              "selector": "target-cluster",
            },
            "name": "Demo Job Template",
            "secret": "grc-testing",
            "type": "AnsibleJob",
          },
          "eventHook": "non-compliance",
          "mode": "once",
          "policyRef": "policy-grc-111",
        },
      },
    ],
  },
}
`;

exports[`Ansible Automation Resolver Correctly resolves ansible credentials 1`] = `
Object {
  "data": Object {
    "ansibleCredentials": Array [
      Object {
        "host": "https://ansible-tower-web-svc-tower.apps.policy-grc-cp-dev-z4bg2.dev08.red-chesterfield.com",
        "name": "toweraccess",
        "namespace": "default",
        "token": "cjRqWnl5eWh1M2pYMzNuUFNrY25oQ2VKQ1ZNQlNN",
      },
      Object {
        "host": "https://ansible-tower-web-svc-tower.apps.policy-grc-cp-dev-z4bg2.dev08.red-chesterfield.com",
        "name": "toweraccess",
        "namespace": "kube-system",
        "token": "cjRqWnl5eWh1M2pYMzNuUFNrY25oQ2VKQ1ZNQlNN",
      },
      Object {
        "host": "https://ansible-tower-web-svc-tower.apps.policy-grc-cp-dev-z4bg2.dev08.red-chesterfield.com",
        "name": "toweraccess",
        "namespace": "policy-namespace",
        "token": "cjRqWnl5eWh1M2pYMzNuUFNrY25oQ2VKQ1ZNQlNN",
      },
      Object {
        "host": "https://ansible-tower-web-svc-tower.apps.policy-grc-cp-dev-z4bg2.dev08.red-chesterfield.com",
        "name": "toweraccess",
        "namespace": "cluster1",
        "token": "cjRqWnl5eWh1M2pYMzNuUFNrY25oQ2VKQ1ZNQlNN",
      },
      Object {
        "host": "https://ansible-tower-web-svc-tower.apps.policy-grc-cp-dev-z4bg2.dev08.red-chesterfield.com",
        "name": "toweraccess",
        "namespace": "local-cluster",
        "token": "cjRqWnl5eWh1M2pYMzNuUFNrY25oQ2VKQ1ZNQlNN",
      },
    ],
  },
}
`;

exports[`Ansible Automation Resolver Should copy and return secret name as credential needs to be copied 1`] = `
Object {
  "data": Object {
    "copyAnsibleSecret": Object {
      "name": "default.toweraccess",
    },
  },
}
`;

exports[`Ansible Automation Resolver Should correctly resolve ansible automation history 1`] = `
Object {
  "data": Object {
    "ansibleAutomationHistories": Array [
      Object {
        "finished": "2021-05-03T17:32:06.357295Z",
        "job": "default/policy-role-once-7fg9p",
        "name": "policy-role-once-7fg9p",
        "namespace": "default",
        "started": "2021-05-03T17:31:59.779211Z",
        "status": "successful",
      },
    ],
  },
}
`;

exports[`Ansible Automation Resolver Should return existing secret name as the copied secret already exists 1`] = `
Object {
  "data": Object {
    "copyAnsibleSecret": Object {
      "name": "existing_toweraccess",
    },
  },
}
`;

exports[`Ansible Automation Resolver Should return root ansible credential name as the targeNamespace is same as credential namespace 1`] = `
Object {
  "data": Object {
    "copyAnsibleSecret": Object {
      "name": "toweraccess",
    },
  },
}
`;

exports[`Correctly Resolves Create and Update Ansible Jobs Mutation 1`] = `
Object {
  "data": Object {
    "createAndUpdateAnsibleJobs": Object {
      "create": Object {
        "errors": Array [],
        "result": Array [
          Object {
            "kind": "PolicyAutomation",
            "response": Object {
              "kind": "PolicyAutomation",
            },
          },
        ],
      },
      "update": Object {
        "errors": Array [],
        "result": Array [
          Object {
            "kind": "PolicyAutomation",
            "response": Object {
              "kind": "PolicyAutomation",
            },
          },
        ],
      },
    },
  },
}
`;
